server/client.go

```diff
@@ -2231,6 +2246,14 @@ func (c *client) processMsgResults(acc *Account, r *SublistResult, msg, subject,
 			continue
 		} else if sub.client.kind == GATEWAY {
 			// Never send to gateway from here.
+			if c.kind == GATEWAY {
+				continue
+			}
+			// The only case we can be here is if this message
```

Comment:
All the special cases means either impl can be simplified or original design is lacking. Not sure which just yet.

Reply:
Removed that, do the check now after this call in processClientInboundMsg.